### PR TITLE
MF-1240 - Return to service transport layer only service errors

### LINKED ISSUE #1240: Return to service transport layer only service errors
Currently, we send to transport.go errors from different sources (db, cache, etc.). However, not all of these errors are addressed by the switch case in the transport.go. So we have a double problem: firstly, inconsistent error sources, secondly unaddressed errors.

Also, we need to make swagger spec in agreement with new error handling, plus, we need to migrate swaggers to openapi 3. 

This relates also to issue #1261, so the single PR should tackle both issues at once.

### PR DESCRIPTION
**Pull request by manuio**
Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

### What does this do?
Refactor encodeError method in services transport layer to ensure that service and db errors are properly encoded in the HTTP response. 

### Which issue(s) does this PR fix/relate to?
Resolves #1240 

### Have you included tests for your changes?
yes

### CODE REVIEW

**Review by blokovi — DISMISSED** [review]

File: auth/api/http/groups/transport.go [hunk 1]
```diff
@@ -327,19 +326,22 @@ func encodeError(_ context.Context, err error, w http.ResponseWriter) {
 		w.WriteHeader(http.StatusForbidden)
 	case errors.Contains(err, auth.ErrMemberAlreadyAssigned):
 		w.WriteHeader(http.StatusConflict)
-	case errors.Contains(err, io.EOF):
-		w.WriteHeader(http.StatusBadRequest)
-	case errors.Contains(err, io.ErrUnexpectedEOF):
-		w.WriteHeader(http.StatusBadRequest)
```
Comment (dborovcanin) [inline auth/api/http/groups/transport.go:333]: Please doublecheck. These errors are not returned by the service layer but can happen during the decoding of the request body.

File: certs/api/transport.go [hunk 2]
```diff
@@ -136,23 +135,27 @@ func decodeRevokeCerts(_ context.Context, r *http.Request) (interface{}, error)
...
+		errors.Contains(err, errors.ErrRemoveEntity):
+		w.WriteHeader(http.StatusInternalServerError)
+
 	default:
-		switch err.(type) {
-		case *json.SyntaxError:
-			w.WriteHeader(http.StatusBadRequest)
-		case *json.UnmarshalTypeError:
-			w.WriteHeader(http.StatusBadRequest)
-		default:
```
Comment (dborovcanin) [inline certs/api/transport.go:155]: Same here, this can happen on the transport layer, not the service.

File: http/api/endpoint_test.go [hunk 3]
```diff
@@ -61,10 +61,10 @@ func (tr testRequest) make() (*http.Response, error) {
 func TestPublish(t *testing.T) {
 	chanID := "1"
 	contentType := "application/senml+json"
-	token := "auth_token"
-	invalidToken := "invalid_token"
+	thinKey := "thing_key"
```
Comment (dborovcanin) [inline http/api/endpoint_test.go]: Fix typo.

File: http/mocks/things.go [hunk 4]
```diff
@@ -31,13 +28,6 @@ func NewThingsClient(data map[string]string) mainflux.ThingsServiceClient {
 func (tc thingsClient) CanAccessByKey(ctx context.Context, req *mainflux.AccessByKeyReq, opts ...grpc.CallOption) (*mainflux.ThingID, error) {
 	key := req.GetToken()
 
-	// Since there is no appropriate way to simulate internal server error,
-	// we had to use this obscure approach. ErrorToken simulates gRPC
-	// call which returns internal server error.
-	if key == ServiceErrToken {
-		return nil, status.Error(codes.Internal, "internal server error")
-	}
-
```
Comment (dborovcanin) [inline http/mocks/things.go]: I'd like to keep both empty and invalid key cases.

File: provision/api/transport.go [hunk 5]
```diff
@@ -89,24 +89,23 @@ func decodeMappingRequest(_ context.Context, r *http.Request) (interface{}, erro
...
+	case errors.Contains(err, errors.ErrAuthentication):
 		w.WriteHeader(http.StatusUnauthorized)
+
 	default:
-		switch err.(type) {
-		case *json.SyntaxError:
-			w.WriteHeader(http.StatusBadRequest)
-		case *json.UnmarshalTypeError:
-			w.WriteHeader(http.StatusBadRequest)
-		default:
```
Comment (dborovcanin) [inline provision/api/transport.go:109]: The same remark: can these errors happen on the transport layer? Check all the other places this section is removed.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: certs/api/transport.go (hunk 2 above)
Comment (manuio) [reply certs/api/transport.go:155]: This errors are related to json.Unmarshal. You can check here https://pkg.go.dev/encoding/json#SyntaxError and here https://pkg.go.dev/encoding/json#UnmarshalTypeError

We are always returning `ErrMalformedEntity` wrapped with the json package error if an error occurs during Unmarshal. This check would be only useful if we return the json error without any wrap.

File: auth/api/http/groups/transport.go (hunk 1 above)
Comment (manuio) [reply auth/api/http/groups/transport.go:333]: This errors are related to protobuf Unmarshal. If there is any error returned we should Wrap it with `ErrMalformedEntity`.  If it occurs in a consumer handler then tthe error is simply never returned.

File: http/api/endpoint_test.go (hunk 3 above)
Comment (manuio) [reply http/api/endpoint_test.go]: thanks

File: http/mocks/things.go (hunk 4 above)
Comment (manuio) [reply http/mocks/things.go]: ok

**Review by dborovcanin — APPROVED** [review]

**Review by blokovi — APPROVED** [review]

File: bootstrap/api/requests.go [hunk 6]
```diff
@@ -8,6 +8,8 @@ import (
 	"github.com/mainflux/mainflux/pkg/errors"
 )
 
+const maxLimitSize = 100
```
Comment (drasko) [inline bootstrap/api/requests.go:11]: No need for this "Size" I think. `maxLimit` is sufficient. Is there some particular reason you introduced this `Size` suffix?

File: bootstrap/api/requests_test.go [hunk 7]
```diff
@@ -198,15 +198,15 @@ func TestListReqValidation(t *testing.T) {
 			desc:   "too large limit",
 			key:    "key",
 			offset: 0,
-			limit:  maxLimit + 1,
+			limit:  maxLimitSize + 1,
```
Comment (drasko) [inline bootstrap/api/requests_test.go:201]: Same here.

**Review by drasko — CHANGES_REQUESTED** [review]

File: bootstrap/api/requests.go (hunk 6 above)
Comment (manuio) [reply bootstrap/api/requests.go:11]: @drasko Consistency reasons. Also think about `maxNameSize` or any other. Are we going to use `maxName`?

Comment (drasko) [reply bootstrap/api/requests.go:11]: Uh... I think it is not exactly the same, but I agree, I would probably keep `maxLimitSize` then. @dusanb94 what do you say?

Comment (dborovcanin) [reply bootstrap/api/requests.go:11]: Keeping it makes sense, since Limit is actually name of the param (though limitSize sounds obnoxious).

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**